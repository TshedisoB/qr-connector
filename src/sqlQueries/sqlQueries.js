export const createTable = (db) => {
  db.transaction((tx) => {
    tx.executeSql(`
      CREATE TABLE IF NOT EXISTS Contacts (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        email TEXT,
        name TEXT,
        profilePicLink TEXT,
        status TEXT,
        website TEXT,
        isFavorite BOOLEAN,
        notes TEXT
      )
    `);
  });
};

export const insertToDataBase = (db, data) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT id FROM Contacts WHERE email = ?;",
      [data.email],
      (_, resultSet) => {
        if (resultSet.rows.length === 0) {
          tx.executeSql(
            "INSERT INTO Contacts (email, name, profilePicLink, status, website, isFavorite, notes) VALUES (?, ?, ?, ?, ?, ?, ?)",
            [
              data.email,
              data.name,
              data.profilePicLink,
              data.status,
              data.website,
              data.isFavorite,
              data.notes,
            ]
          );
        }
      }
    );
  });
};

export const getAllContacts = (db, callback) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT * FROM Contacts ORDER BY isFavorite DESC",
      [],
      (_, resultSet) => {
        const contacts = resultSet.rows;
        const contactsArray = [];
        for (let i = 0; i < contacts.length; i++) {
          contactsArray.push(contacts.item(i));
        }
        callback(contactsArray);
      },
      () => {
        callback([]);
      }
    );
  });
};

export const updateFavorites = (db, email, isFavorite) => {
  db.transaction((tx) => {
    tx.executeSql("UPDATE Contacts SET isFavorite = ? WHERE email = ?;", [
      !isFavorite,
      email,
    ]);
  });
};

export const deleteContact = (db, email, callback) => {
  db.transaction((tx) => {
    tx.executeSql(
      "DELETE FROM Contacts WHERE email = ?;",
      [email],
      (_, resultSet) => {
        const rowsAffected = resultSet.rowsAffected;
        callback(rowsAffected > 0);
      },
      () => {
        callback(false);
      }
    );
  });
};

export const addOrUpdateNotes = (db, email, notes) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT * FROM Contacts WHERE email = ?;",
      [email],
      (_, resultSet) => {
        const existingContact = resultSet.rows.item(0);

        if (existingContact) {
          const updatedContact = { ...existingContact, notes };

          tx.executeSql("UPDATE Contacts SET notes = ? WHERE email = ?;", [
            updatedContact.notes,
            updatedContact.email,
          ]);
        }
      }
    );
  });
};
