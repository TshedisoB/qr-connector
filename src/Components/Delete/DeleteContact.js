import React, { useState } from "react";
import {
  View,
  Text,
  TouchableOpacity,
  Modal,
  Button,
  Alert,
} from "react-native";
import { useNavigation } from "@react-navigation/native";
import { AntDesign } from "@expo/vector-icons";

import { deleteContact } from "../../sqlQueries/sqlQueries";
import { styles } from "./DeleteContact.style";
import { db } from "../ScanQr/ScanQr";

function DeleteContact({ contact }) {
  const navigation = useNavigation();
  const [isModalVisible, setModalVisible] = useState(false);

  const handleDeleteContact = () => {
    setModalVisible(true);
  };

  const handleConfirmDelete = () => {
    const email = contact[0].email;

    deleteContact(db, email, (isDeleted) => {
      if (isDeleted) {
        Alert.alert(
          "Success",
          `Contact '${contact[0].name}' deleted successfully.`
        );
        navigation.navigate("Contact List");
      } else {
        Alert.alert(
          "Error",
          `Failed to delete contact '${contact[0].name}'. Please try again.`
        );
      }
    });
    setModalVisible(false);
  };

  const handleCancelDelete = () => {
    setModalVisible(false);
  };

  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={styles.deleteText}
        onPress={handleDeleteContact}
        testID="delete-contact">
        <Text>Delete</Text>
        <Text>Contact</Text>
        <AntDesign name="delete" style={styles.antIcon} />
      </TouchableOpacity>

      <Modal
        animationType="slide"
        transparent={true}
        visible={isModalVisible}
        onRequestClose={() => setModalVisible(false)}>
        <View style={styles.modalContainer}>
          <View style={styles.modalContent}>
            <Text>Are you sure you want to delete this contact?</Text>
            <View style={styles.deleteButtons}>
              <Button
                onPress={handleCancelDelete}
                title="Cancel"
                color="grey"
                testID="cancel-delete"
              />
              <Button
                onPress={handleConfirmDelete}
                title="Delete"
                color="red"
                testID="confirm-delete"
              />
            </View>
          </View>
        </View>
      </Modal>
    </View>
  );
}

export default DeleteContact;
