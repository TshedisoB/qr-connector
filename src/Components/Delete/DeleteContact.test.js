import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { render, fireEvent, waitFor } from "@testing-library/react-native";

import * as sqlQueries from "../../sqlQueries/sqlQueries";
import DeleteContact from "./DeleteContact";

jest.mock("../../sqlQueries/sqlQueries", () => ({
  __esModule: true,
  deleteContact: jest.fn(),
}));

const contactData = [
  {
    name: "Tshediso Boshiana",
    email: "tshediso.boshiana@umuzi.org",
  },
];

describe("DeleteContact", () => {
  let tree;

  beforeEach(() => {
    tree = render(
      <NavigationContainer>
        <DeleteContact contact={contactData} />
      </NavigationContainer>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render the component correctly", () => {
    expect(tree).toMatchSnapshot();
  });

  it("should display the prompt when 'Delete Contact' trash icon is pressed", () => {
    const { getByTestId, getByText } = tree;

    fireEvent.press(getByTestId("delete-contact"));

    expect(
      getByText("Are you sure you want to delete this contact?")
    ).toBeDefined();
  });

  it("should `delete` a contact when `Delete` is button is pressed after prompting.", async () => {
    const { getByTestId, getByText } = tree;

    fireEvent.press(getByTestId("delete-contact"));

    expect(
      getByText("Are you sure you want to delete this contact?")
    ).toBeDefined();

    fireEvent.press(getByTestId("confirm-delete"));

    await waitFor(() => {
      expect(sqlQueries.deleteContact).toHaveBeenCalledWith(
        expect.anything(),
        expect.anything(),
        expect.anything()
      );
    });
  });

  it("should cancel contact deletion when 'Cancel' is pressed", () => {
    const { getByTestId, getByText } = tree;

    fireEvent.press(getByTestId("delete-contact"));

    expect(
      getByText("Are you sure you want to delete this contact?")
    ).toBeDefined();

    fireEvent.press(getByTestId("cancel-delete"));

    expect(sqlQueries.deleteContact).not.toHaveBeenCalled();
  });
});
