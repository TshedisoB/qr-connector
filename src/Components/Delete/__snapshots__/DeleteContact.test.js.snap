// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteContact should render the component correctly 1`] = `
<View
  style={
    {
      "alignItems": "flex-end",
      "flex": 1,
      "marginRight": 20,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#eee",
        "borderRadius": 5,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="delete-contact"
  >
    <Text>
      Delete
    </Text>
    <Text>
      Contact
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 12,
          },
          {
            "color": "green",
            "fontSize": 22,
          },
          {
            "fontFamily": "anticon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      îšŸ
    </Text>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
